### PR TITLE
Telemetry state change event

### LINKED ISSUE #6032: Add telemetry state change event
Adding this stub in reference to internal issue and discussion here: https://github.com/meltano/internal-general/issues/338

This will give better insight into users stopping using meltano (or changing their project IDs) vs choosing to opt out of anonymous usage stats.

### PR DESCRIPTION
**Pull request by aaronsteers**
Resolves https://github.com/meltano/internal-general/issues/338

This adds a one-time 'telemetry_state_change' event when the value of either project_id or send_anonymous_usage_stats changes.

~~Note: this borrows from #5941 and targets that branch.~~

~~Plan is that #5941 will merge first and then this will be updated to target `main`.~~ Done.

TODO:

- [x] Feature development
- [x] Define Snowplow schema for the new event type
- [ ] Add tests
- [x] Manual sanity check
- [x] Register the new event type with Snowcatcloud

### CODE REVIEW

File: src/meltano/core/tracking/tracker.py
Comment (edgarrmondragon) [inline src/meltano/core/tracking/tracker.py:54]: Wdyt about calling this class `Telemetry settings`?

File: src/meltano/core/tracking/tracker.py
Comment (edgarrmondragon) [inline src/meltano/core/tracking/tracker.py:310]: IMO it's a bit weird to have a "private" computed property. What do you think of making this a proper method since it's only called once anyway?

@WillDaSilva or @pandemicsyn might have opinions too 🙂

File: src/meltano/core/tracking/tracker.py
Comment (pandemicsyn) [inline src/meltano/core/tracking/tracker.py:12]: @aaronsteers  that from __future__ import annotations allows new style annotations, and we've been slowly moving the code base to those. That allows referring list , tuple, and friends without having to import anything from the typing package.

If you're running the pre-commit hook pyupgrade will catch a lot of these for you, but figured I'd give a ya heads up.

File: src/meltano/core/tracking/tracker.py
Comment (pandemicsyn) [inline src/meltano/core/tracking/tracker.py:60]: for ref, these can convert to:

```
client_id: uuid.UUID | None
```

File: src/meltano/core/tracking/tracker.py
Comment (pandemicsyn) [reply src/meltano/core/tracking/tracker.py:310]: +1 I'd prefer this one be a method, IMHO this does a lot for being a "property" .

File: src/meltano/core/tracking/tracker.py
Comment (pandemicsyn) [inline src/meltano/core/tracking/tracker.py:319]: can we use a structured log entry here. Something like:

```logger.warn("Unexpected analytics.json setting type", from_val=from_val, current_type=type(from_val))```

File: src/meltano/core/tracking/tracker.py
Comment (pandemicsyn) [inline src/meltano/core/tracking/tracker.py:332]: Ditto the structlog entry here would be nice.

**Comment by edgarrmondragon** [comment]
The commit history in this one has things it shouldn't, right? Does it need to be rebased?

File: src/meltano/core/tracking/tracker.py
Comment (aaronsteers) [reply src/meltano/core/tracking/tracker.py:54]: Sure! Initially it was named after the analytics.json file, but TelemetrySettings seems more appropriate, I think. 👍

File: src/meltano/core/tracking/tracker.py
Comment (aaronsteers) [reply src/meltano/core/tracking/tracker.py:310]: Agreed - this one grew and morphed quite a lot. I tidied it and moved the helper function out to its own private method.

File: src/meltano/core/tracking/tracker.py
Comment (aaronsteers) [reply src/meltano/core/tracking/tracker.py:12]: Updated per your suggestion. Thanks!

I considered putting them in the new style before, but my IDE linter is complaining about the new types. Going forward, I may need to find a fix for the mypy lint issue, perhaps using a higher Python version for my development, since it seems mypy prefers 3.9 or 3.10.

**Comment by pandemicsyn** [comment]
@WillDaSilva heads up on https://github.com/meltano/meltano/pull/6028 which should hopefully land this morning, it'll probably give you some merge conflicts.

**Comment by WillDaSilva** [comment]
@aaronsteers @pandemicsyn @pnadolny13 The new event type `telemetry_state_change_event` has been added to Snowcatcloud. Pending your approval, I'll put it into production mode.

**Comment by pandemicsyn** [comment]
@WillDaSilva fyi - https://github.com/meltano/meltano/pull/6028 landed so probably have a little merge/commit clean up.

**Comment by WillDaSilva** [comment]
@aaronsteers I am unable to add you as a reviewer.

**Comment by aaronsteers** [comment]
@WillDaSilva - Because I authored it I guess I cannot be an approver. I'll review now and add any comments I see.

File: src/meltano/core/tracking/tracker.py
Comment (aaronsteers) [inline src/meltano/core/tracking/tracker.py:150]: Yes, this is the bit that resolves the removal of the `env_alias` `MELTANO_TRACKING_DISABLED`.

**Review by aaronsteers — COMMENTED** [review]
Looks good! ✅

**Review by pandemicsyn — APPROVED** [review]
lgtm! Also double checked,  I think i see the correct # of events at the correct times, in snowplow micro

**Pull request merged**
